File: code/Android_App/ITS-Mobile-App/app/src/main/java/pt/isel/meic/tfm/itsmobileapp/security_protocols/ProtocolSetup.kt
package pt.isel.meic.tfm.itsmobileapp.security_protocols

import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.DLAPP
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.biometric_device.BiometricDeviceImpl
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.biometric_device.IBiometricDevice
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.tamper_proof_device.ITamperProofDeviceDLAPP
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.tamper_proof_device.TamperProofDeviceDLAPPImpl
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.tamper_proof_device.hash_chain.HashChainModuleImpl
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.tamper_proof_device.pseudo_identity.PseudoIdentityModuleImpl
import pt.isel.meic.tfm.itsmobileapp.security_protocols.mfspv.MFSPV
import pt.isel.meic.tfm.itsmobileapp.security_protocols.mfspv.tamper_proof_device.ITamperProofDeviceMFSPV
import pt.isel.meic.tfm.itsmobileapp.security_protocols.mfspv.tamper_proof_device.TamperProofDeviceMFSPVImpl
import pt.isel.meic.tfm.itsmobileapp.security_protocols.mfspv.tamper_proof_device.pseudo_identity.IPseudoIdentityModMFSPV
import pt.isel.meic.tfm.itsmobileapp.security_protocols.mfspv.tamper_proof_device.pseudo_identity.PseudoIdentityModMFSPVImpl
import pt.isel.meic.tfm.itsmobileapp.security_protocols.no_security.NoSecurity

class ProtocolSetup {

    companion object {

        fun getSecurityProtocol(protocol: Common.ProtocolEnum): ISecurityProtocol {
            return when (protocol) {
                Common.ProtocolEnum.DLAPP -> dlappSetup()
                Common.ProtocolEnum.MFSPV -> mfspvSetup()
                Common.ProtocolEnum.NONE -> noSecuritySetup()
            }
        }

        /**
         * No security
         */
        private fun noSecuritySetup() : NoSecurity {
            return NoSecurity()
        }

        /**
         * DLAPP protocol configuration and initialisation
         */
        private fun dlappSetup() : DLAPP {

            // Define algorithm arguments
            val secretSystemKey : ByteArray = byteArrayOf(-17, -12, -59, -126, -37, -128, -5, 80, -30, -36, 27, -82, -113, 69, 60, 80, -64, 65, 99, -44, 99, 124, 105, -38, -95, 25, -7, 19, 38, 45, 38, 45)
            val hashChainElements : Int = 5
            val numberOfPseudoIdentities : Int = 3
            val hashAlgorithm : String = "SHA-256"
            val signatureAlgorithm : String = "HmacSHA256"

            // Create an instance of type ITamperProofDevice
            val tpd : ITamperProofDeviceDLAPP = TamperProofDeviceDLAPPImpl(
                hashChainModule = HashChainModuleImpl(
                    hashAlgorithm = hashAlgorithm
                ),
                pseudoIdentityModule = PseudoIdentityModuleImpl(),
                secretSystemKey = secretSystemKey,
                numberOfHashChainElements = hashChainElements,
                numberOfPseudoIdentities = numberOfPseudoIdentities,
                macAlgorithm = signatureAlgorithm,
            )

            // Create an instance of type IBiometricDevice
            val bd : IBiometricDevice = BiometricDeviceImpl()

            return DLAPP(
                biometricDevice = bd,
                tamperProofDevice = tpd
            )

        }

        /**
         * MFSPV protocol configuration and initialisation
         */
        private fun mfspvSetup() : MFSPV {
            val pidGenerator : IPseudoIdentityModMFSPV = PseudoIdentityModMFSPVImpl(
                hashAlgorithm = "SHA-256",
                apiNew = byteArrayOf(
                    -6, -6, 6, -1, -3, -1, -51, 81, -71, 1, -71, 1,
                    -6, -6, 6, -1, -3, -1, -51, 81, -71, 1,
                    -6, -6, 6, -1, -3, -1, -51, 81, -71, 1
                ),

                idV = byteArrayOf(
                    -10, -20, -30, -40, -50, -60, -70, -80, -90, -10, -90, -10,
                    -10, -20, -30, -40, -50, -60, -70, -80, -90, -10,
                    -10, -20, -30, -40, -50, -60, -70, -80, -90, -10
                ),

                vSk = byteArrayOf(
                    -60, -62, 60, -10, -37, -12, -5, 8, -7, 19, -7, 19,
                    -60, -62, 60, -10, -37, -12, -5, 8, -7, 19,
                    -60, -62, 60, -10, -37, -12, -5, 8, -7, 19
                ),

                kMbr = byteArrayOf(
                    -64, -65, 4, -40, -7, -2, 5, -8, -11, 22, -11, 22,
                    -64, -65, 4, -40, -7, -2, 5, -8, -11, 22,
                    -64, -65, 4, -40, -7, -2, 5, -8, -11, 22
                )
            )

            val tpd : ITamperProofDeviceMFSPV = TamperProofDeviceMFSPVImpl(
                pseudoIdentityModule = pidGenerator,
                // Region's Key (32 bytes)
                regionalKey = byteArrayOf(
                    -16, 9, -52, -51, -42, -96, 104, 101, 7, 17, -91, 42,
                    -72, 56, 112, 76, 103, 115, -19, -33, 31, -119,
                    82, 113, -94, 114, -112, -81, -57, -94, 106, -40),
                hashAlgorithm = "SHA-256"
            )

            return MFSPV(tamperProofDevice = tpd)
        }

    }

}

File: code/MRSU_App/security_protocols/protocol_setup.py
import logging

from security_protocols.common import SecurityProtocolEnum
from security_protocols.dlapp.dlapp_impl import DLAPP
from security_protocols.dlapp.biometric_device.biometric_device_impl import BiometricDeviceImpl
from security_protocols.dlapp.tamper_proof_device.pseudo_identity_module_impl import PseudoIdentityModuleImpl
from security_protocols.dlapp.tamper_proof_device.hash_chain_module_impl import HashChainModuleImpl
from security_protocols.dlapp.tamper_proof_device.tamper_proof_device_impl import TamperProofDeviceDLAPPImpl
from common.util import create_signed_byte_array_from
from security_protocols.mfspv.mfspv_impl import MFSPV
from security_protocols.mfspv.tamper_proof_device.pid_generator_impl import PseudoIdentityMFSPVImpl
from security_protocols.mfspv.tamper_proof_device.tamper_proof_device_impl import TamperProofDeviceMFSPVImpl
from security_protocols.no_security.no_security_impl import NoSecurity


class ProtocolSetup:

    @staticmethod
    def get_security_protocol(security_prot):
        logging.debug(f'get_security_protocol called with security_prot = {security_prot}')

        if security_prot == SecurityProtocolEnum.NONE:
            return ProtocolSetup.no_security_setup()

        elif security_prot == SecurityProtocolEnum.MFSPV:
            return ProtocolSetup.mfspv_setup()

        elif security_prot == SecurityProtocolEnum.DLAPP:
            return ProtocolSetup.dlapp_setup()

    @staticmethod
    def no_security_setup():
        """
        NoSecurity configuration
        @return: no security object
        """
        logging.info(f'No security setup')

        return NoSecurity()

    @staticmethod
    def dlapp_setup():
        """
        DLAPP configuration
        @return: security protocol object with DLAPP configuration
        """
        logging.info(f'DLAPP protocol setup')

        # ---------------------------- #
        # ------ dlapp Protocol ------ #
        # ---------------------------- #

        secret_system_key = create_signed_byte_array_from(byte_array=[-17, -12, -59, -126, -37, -128, -5, 80,
                                                                      -30, -36, 27, -82, -113, 69, 60, 80, -64,
                                                                      65, 99, -44, 99, 124, 105, -38, -95, 25,
                                                                      -7, 19, 38, 45, 38, 45])

        number_of_hash_chain_elements = 5
        number_of_pseudo_identities = 3
        hash_algorith = "sha256"

        pseudo_id_module = PseudoIdentityModuleImpl()
        hash_chain_module = HashChainModuleImpl(hash_algorithm=hash_algorith)

        tp_device = TamperProofDeviceDLAPPImpl(
            hash_chain_module=hash_chain_module,
            pseudo_identity_module=pseudo_id_module,
            secret_system_key=secret_system_key,
            number_of_hash_chain_elements=number_of_hash_chain_elements,
            number_of_pseudo_identities=number_of_pseudo_identities
        )
        biometric_device = BiometricDeviceImpl(return_value=True)

        return DLAPP(
            biometric_device=biometric_device,
            tamper_proof_device=tp_device)

    @staticmethod
    def mfspv_setup():
        """
        MFSPV configuration
        @return: security protocol object with MFSPV configuration
        """

        pid_module = PseudoIdentityMFSPVImpl(
            hash_algorithm="sha256",

            # ITS-S specific info:
            api_new=create_signed_byte_array_from(byte_array=[
                -17, -19, -9, -120, -37, -12, -5, 80, -7, 19, -7, 19,
                -17, -19, -9, -120, -37, -12, -5, 80, -7, 19,
                -17, -19, -9, -120, -37, -12, -5, 80, -7, 19
            ]),

            v_sk=create_signed_byte_array_from(byte_array=[
                -11, -12, -13, -14, -15, -16, -17, -18, -19, -20, -19, -20,
                -11, -12, -13, -14, -15, -16, -17, -18, -19, -20,
                -11, -12, -13, -14, -15, -16, -17, -18, -19, -20
            ]),

            id_v=create_signed_byte_array_from(byte_array=[
                -1, -2, -3, -4, -5, -6, -7, -8, -9, -10, -9, -10,
                -1, -2, -3, -4, -5, -6, -7, -8, -9, -10,
                -1, -2, -3, -4, -5, -6, -7, -8, -9, -10
            ]),

            k_mbr=create_signed_byte_array_from(byte_array=[
                -11, -19, -9, -120, -37, -12, -5, 80, -7, 19, -7, 19,
                -11, -19, -9, -120, -37, -12, -5, 80, -7, 19,
                -11, -19, -9, -120, -37, -12, -5, 80, -7, 19
            ])

        )

        tp_device = TamperProofDeviceMFSPVImpl(
            pseudo_identity_module=pid_module,
            # Region's Key (32 bytes)
            regional_key=create_signed_byte_array_from(byte_array=[
                -16, 9, -52, -51, -42, -96, 104, 101, 7, 17, -91, 42,
                -72, 56, 112, 76, 103, 115, -19, -33, 31, -119,
                82, 113, -94, 114, -112, -81, -57, -94, 106, -40
            ]),
            hash_algorithm="sha256"
        )

        return MFSPV(tamper_proof_device=tp_device)
